File: blaze_flutter_sdk/android/src/main/kotlin/com/blaze/flutterblazesdk/widgets/customization/WidgetLayoutParser.kt
package com.blaze.flutterblazesdk.widgets.customization

import android.content.Context
import com.blaze.blazesdk.style.widgets.BlazeWidgetLayout
import com.blaze.flutterblazesdk.utils.parsing.gson.fromJsonString
import com.blaze.flutterblazesdk.utils.parsing.gson.toObject

/**
 * Extension function to convert a Map to BlazeReactWidgetLayout using custom Gson with adapters
 */
fun Map<String, Any?>.toBlazeReactWidgetLayout(): BlazeReactWidgetLayout? {
    return this.toObject<BlazeReactWidgetLayout>()
}

/**
 * Extension function to apply widget layout customization over a preset layout
 */
fun applyWidgetLayoutCustomization(
        reactWidgetLayout: Map<String, Any?>?,
        widgetLayout: BlazeWidgetLayout,
        context: Context
): BlazeWidgetLayout {
    val rawLayout = reactWidgetLayout ?: return widgetLayout

    val customization = rawLayout.toBlazeReactWidgetLayout()

    return widgetLayout.mergedWith(customization = customization, context)
}

/**
 * Transforms the map into a Map of BlazeReactWidgetItemCustomMapping to
 * BlazeReactWidgetItemStyleOverrides
 */
fun Map<String, Any>.toReactWidgetStylesOverridesMap():
        Map<BlazeReactWidgetItemCustomMapping, BlazeReactWidgetItemStyleOverrides>? {
    val resultMap =
            mutableMapOf<BlazeReactWidgetItemCustomMapping, BlazeReactWidgetItemStyleOverrides>()

    // Iterate through each key-value pair in the map
    for ((key, value) in this) {
        // Attempt to decode the key into BlazeReactWidgetItemCustomMapping
        val keyValue = key.fromJsonString<BlazeReactWidgetItemCustomMapping>() ?: continue

        // Attempt to convert the value to BlazeReactWidgetItemStyleOverrides
        val valueValue =
                (value as? Map<String, Any>)?.toObject<BlazeReactWidgetItemStyleOverrides>()
                        ?: continue

        // Add to the result map
        resultMap[keyValue] = valueValue
    }

    return if (resultMap.isEmpty()) null else resultMap
}

File: blaze_flutter_sdk/ios/Sources/Widgets/LayoutCustomization/WidgetLayoutParser.swift
import BlazeSDK
import Foundation

extension Dictionary where Key == String, Value == AnyHashable {
    /**
     * Extension function to convert a Dictionary to BlazeReactWidgetLayout using JSONDecoder
     */
    func toBlazeReactWidgetLayout() -> BlazeReactWidgetLayout? {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: self, options: [])
            let decoder = JSONDecoder()
            return try decoder.decode(BlazeReactWidgetLayout.self, from: jsonData)
        } catch {
            return nil
        }
    }

    /**
     * Extension function to convert a Dictionary to BlazeReactWidgetStylesOverrides using JSONDecoder
     */
    var toReactWidgetStylesOverrides: BlazeReactWidgetStylesOverrides? {
        var result: BlazeReactWidgetStylesOverrides?

        do {
            let jsonData = try JSONSerialization.data(withJSONObject: self, options: [])
            result = try JSONDecoder().decode(BlazeReactWidgetStylesOverrides.self, from: jsonData)
        } catch {
            print("Error decoding BlazeReactWidgetStylesOverrides: \(error)")
        }

        return result
    }

    /// Transforms the dictionary into a map of BlazeReactWidgetItemCustomMapping to BlazeReactWidgetStylesOverrides
    var toReactWidgetStylesOverridesMap:
        [BlazeReactWidgetItemCustomMapping: BlazeReactWidgetStylesOverrides]?
    {
        var resultMap = [BlazeReactWidgetItemCustomMapping: BlazeReactWidgetStylesOverrides]()

        // Iterate through each key-value pair in the dictionary
        for (key, value) in self {
            // Attempt to decode the key into BlazeReactWidgetItemCustomMapping
            guard let keyData = key.data(using: .utf8),
                let keyValue = try? JSONDecoder().decode(
                    BlazeReactWidgetItemCustomMapping.self, from: keyData)
            else {
                continue
            }

            // Attempt to serialize the value dictionary and decode it into BlazeReactWidgetStylesOverrides
            guard let valueDict = value as? [String: AnyHashable],
                let valueData = try? JSONSerialization.data(withJSONObject: valueDict, options: []),
                let valueValue = try? JSONDecoder().decode(
                    BlazeReactWidgetStylesOverrides.self, from: valueData)
            else {
                continue
            }

            // Add to the result map
            resultMap[keyValue] = valueValue
        }

        return resultMap.isEmpty ? nil : resultMap
    }
}
